### ISSUE TITLE
Allow inline method chaining if line length does not exceed `printWidth`

### ISSUE THREAD

**Issue by sharmilajesupaul** [prettier/prettier#3107]
Just to give you some context on this, I've been working with @ljharb to bring prettier to @airbnb. 
We noticed a few cases where Prettier’s formatting could be adjusted to more closely align with our style guide.

cc. @vjeux

**Prettier 1.7.4**

If a method chain has more than 1 method, Prettier creates a linebreak after each chained method, increasing the indentation in the method body.

[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAEnADYkQDqEATiQCYAUAlPsADpT74DuOUdEbgDoAYhAgBhEjgQwhAZzgwAagENq8hu05d85MKpJwk+AOZKxEADIQDRrfgAOq8-gC+TADQdd+VehgACxocAC9VXGgAFQgAawQTOgjVIRg4hG8dDyE+HBU4ahxUAE8GAHJUcQB9SFhVPgKypgBuDh98VDgAK1UADyFVHt6tdq50WhMygHoANwKinANIqCmyzN8kmFUo4sdjfDKu+Wg19uyghAZR-AZN1RYAXgA+VmuuIJx5BSUAZS34ByfGyqOh8UwmVCGRTuFpvfAfL65GAUHB0cwwW7JWE6Lhuda6TFbR4vbS+d6BT5CcimACSAFtHNQ4PJ5HgoOVKhAqp0YGBAlUCtQaGt8HdsWT8CISKp5IEhIKaAwaQBGAAcUFS5QgqFQ0igcCqqNkeWKQk5hpZ6DgTXFujx7WxbhAnhAEEcy3kyFAMpgyEhJEULoARtRVGB4jAfs4wGDkDBqFaXQIwH6oXAXUdegAhUPh36qOlwKwNVMB9OugKOAIAJlLgZAzg0BWQICDqiDxXIUGdDcKsBRdCCyGVAAYRy7FHScHGE+XWVBTEYAIroCDwGeJkBbIMDodIasu+P1PWmCQQOl01QtwwkHvoRRRduepD+xRuNxAA)
```sh
# Options (if any):
--single-quote
--print-width 100
--trailing-comma "all"
```

**Input:**
```jsx
function HelloWorld() {
  window.FooClient.setVars({
    locale: getFooLocale({ page }),
    authorizationToken: data.token,
  }).initVerify('foo_container');

  fejax.ajax({
    url: '/verification/',
    dataType: 'json',
  }).then(
    (data) => {
      this.setState({ isLoading: false });
      this.initWidget(data);
    },
    (data) => {
      this.logImpression('foo_fetch_error', data);
      Flash.error(I18n.t('offline_identity.foo_issue'));
    },
  );
}
```

**Output:**
```jsx
function HelloWorld() {
  window.FooClient
    .setVars({
      locale: getFooLocale({ page }),
      authorizationToken: data.token,
    })
    .initVerify('foo_container');

  fejax
    .ajax({
      url: '/verification/',
      dataType: 'json',
    })
    .then(
      data => {
        this.setState({ isLoading: false });
        this.initWidget(data);
      },
      data => {
        this.logImpression('foo_fetch_error', data);
        Flash.error(I18n.t('offline_identity.foo_issue'));
      },
    );
}
```

**Desired behavior:**
Creating a line break after chained methods can increase the indentation of multiline function bodies. It would help to allow chaining to be inline until it exceeds the `printWidth`. This shouldn't cause a readability issue since multiline blocks will still be encapsulated.

eg: 
```jsx
function HelloWorld() {
  window.FooClient.setVars({
    locale: getFooLocale({ page }),
    authorizationToken: data.token,
  }).initVerify('foo_container');

  fejax.ajax({
    url: '/verification/',
    dataType: 'json',
  }).then(
    (data) => {
      this.setState({ isLoading: false });
      this.initWidget(data);
    },
    (data) => {
      this.logImpression('foo_fetch_error', data);
      Flash.error(I18n.t('offline_identity.foo_issue'));
    },
  );
}
```
-----

Diff between Prettier & desired output:
```diff
*** prettier.js	2017-10-25 16:40:41.000000000 -0700
--- desired.js	2017-10-25 16:40:39.000000000 -0700
***************
*** 1,24 ****
  function HelloWorld() {
!   window.FooClient
!     .setVars({
!       locale: getFooLocale({ page }),
!       authorizationToken: data.token,
!     })
!     .initVerify('foo_container');
  
!   fejax
!     .ajax({
!       url: '/verification/',
!       dataType: 'json',
!     })
!     .then(
!       data => {
!         this.setState({ isLoading: false });
!         this.initWidget(data);
!       },
!       data => {
!         this.logImpression('foo_fetch_error', data);
!         Flash.error(I18n.t('offline_identity.foo_issue'));
!       },
!     );
  }
--- 1,20 ----
  function HelloWorld() {
!   window.FooClient.setVars({
!     locale: getFooLocale({ page }),
!     authorizationToken: data.token,
!   }).initVerify('foo_container');
  
!   fejax.ajax({
!     url: '/verification/',
!     dataType: 'json',
!   }).then(
!     (data) => {
!       this.setState({ isLoading: false });
!       this.initWidget(data);
!     },
!     (data) => {
!       this.logImpression('foo_fetch_error', data);
!       Flash.error(I18n.t('offline_identity.foo_issue'));
!     },
!   );
  }
```

**Comment by vjeux** [comment]
For context, method chaining is the piece of JavaScript that was the most difficult to print, so any non trivial change like the one you are suggesting is likely going to require a big time investment.

From the example, it looks like you are looking to remove all the newlines and indentation. I tried to do that and some of the unit tests look worse. I'd love to know how you would write them:

```diff
-  return action$
-    .ofType(ActionTypes.SEARCHED_USERS)
-    .map(action => action.payload.query)
-    .filter(q => !!q)
-    .switchMap(q =>
-      Observable.timer(800) // debounce
-        .takeUntil(action$.ofType(ActionTypes.CLEARED_SEARCH_RESULTS))
-        .mergeMap(() =>
-          Observable.merge(
-            Observable.of(replace(`?q=${q}`)),
-            ajax
-              .getJSON(`https://api.github.com/search/users?q=${q}`)
-              .map(res => res.items)
-              .map(receiveUsers)
-          )
-        )
-    );
+  return action$.ofType(ActionTypes.SEARCHED_USERS).map(
+    action => action.payload.query
+  ).filter(q => !!q).switchMap(q =>
+    Observable.timer(800).takeUntil( // debounce
+      action$.ofType(ActionTypes.CLEARED_SEARCH_RESULTS)
+    ).mergeMap(() =>
+      Observable.merge(
+        Observable.of(replace(`?q=${q}`)),
+        ajax.getJSON(`https://api.github.com/search/users?q=${q}`).map(
+          res => res.items
+        ).map(receiveUsers)
+      )
+    )
+  );
```

```diff
-Observable.of(process)
-  // Don't complete until we say so!
-  .merge(Observable.never())
-  // Get the errors.
-  .takeUntil(throwOnError ? errors.flatMap(Observable.throw) : errors)
-  .takeUntil(exit);
+Observable.of(process)// Don't complete until we say so!
+.merge(Observable.never())// Get the errors.
+.takeUntil(throwOnError ? errors.flatMap(Observable.throw) : errors).takeUntil(
+  exit
+);
```

```diff
-nock(/test/)
-  .matchHeader("Accept", "application/json")
-  [httpMethodNock(method)]("/foo")
-  .reply(200, {
-    foo: "bar"
-  });
+nock(/test/).matchHeader("Accept", "application/json")[httpMethodNock(method)](
+  "/foo"
+).reply(200, {
+  foo: "bar"
+});
```

```diff
-    })
-      .filter(data => theFilter(data))
-      .map(({ theType, ...data }) => theMap(theType, data))
-      .retryWhen(errors => errors)
+    }).filter(data => theFilter(data)).map(({ theType, ...data }) =>
+      theMap(theType, data)
+    ).retryWhen(errors => errors)
```

If you want to try it for yourself:

```diff
diff --git a/src/printer.js b/src/printer.js
index 42884b55..f4b4951d 100644
--- a/src/printer.js
+++ b/src/printer.js
@@ -3578,6 +3578,7 @@ function printMemberChain(path, options, print) {
     if (groups.length === 0) {
       return "";
     }
+    return group(join("", groups.map(printGroup)));
     return indent(
       group(concat([hardline, join(hardline, groups.map(printGroup))]))
     );
```

**Comment by sharmilajesupaul** [comment]
You're right, those diffs do look worse and the single heuristic: _chain everything, until you hit the end of the line_, is definitely not going to work. The leading and inline comments get completely displaced too.

What if Prettier only changed the first property invocation to be inline?

i.e. the first diff would look like:
```js
return action$.ofType(ActionTypes.SEARCHED_USERS)
  .map(action => action.payload.query)
  .filter(q => !!q)
  .switchMap(q =>
    Observable.timer(800) // debounce
      .takeUntil(action$.ofType(ActionTypes.CLEARED_SEARCH_RESULTS))
      .mergeMap(() =>
        Observable.merge(
          Observable.of(replace(`?q=${q}`)),
          ajax
            .getJSON(`https://api.github.com/search/users?q=${q}`)
            .map(res => res.items)
            .map(receiveUsers)
        )
      )
  );
``` 
and it would mirror the behavior of this Prettier output: 
```js 
  window.FooClient
    .setVars({
      locale: getFooLocale({ page }),
      authorizationToken: data.token,
    })
    .initVerify('foo_container');
```
@ljharb do you agree?

**Comment by ljharb** [comment]
I think that, when the first invocation can fit on one line, this would work well.

However I think the real conceptual difference is that in `window.FooClient.setVars(` or `[].map(`, the invocation is conceptually a separate thing from the receiver - but in `Promise.resolve(),` and `$.ajax(`, the invocation and the receiver are part of the same conceptual thing and thus shouldn't be split apart across lines.

I'm not sure there's a single good heuristic around this that can be programmatically applied while meeting our requirements.

@vjeux would it be possible to have an option to avoid altering method chains, while still styling their arguments and bodies?

**Comment by vjeux** [comment]
Looking at the two examples, we are already doing work to handle this case. The idea as you mentioned is to figure out what is the actual thing that we're working on.

Right now the heuristic for that is:
- It starts with a capital letter
- It is $ or _

If we slightly tweak your inputs:
- window.FooClient -> FooClient
- fejax -> $

then it works as you suggested:

```js
FooClient.setVars({
  locale: getFooLocale({ page }),
  authorizationToken: data.token
}).initVerify("foo_container");

$.ajax({
  url: "/verification/",
  dataType: "json"
}).then(
  data => {
    this.setState({ isLoading: false });
    this.initWidget(data);
  },
  data => {
    this.logImpression("foo_fetch_error", data);
    Flash.error(I18n.t("offline_identity.foo_issue"));
  }
);
```

For the second one, we can improve the heuristic so that `window.FooClient` matches. For the second one I doubt that we can. How often in your codebase do you have `fejax.ajax`. Do you have more examples where the heuristic is failing?

**Comment by vjeux** [comment]
By the way, are those the only 3 issues you are facing at Airbnb or are there a bunch more? Also, how often do those 3 happen in practice on your codebase? This would be very useful to prioritize.

**Comment by vjeux** [comment]
For reference, here's the code that contains the heuristic: https://github.com/prettier/prettier/blob/6ccf5c0246970557426220008d2f950e67101181/src/printer.js#L3781-L3800

**Comment by vjeux** [comment]
I sent #3112 to fix the case with `window.FooClient.setVars `

**Comment by sharmilajesupaul** [comment]
@vjeux To answer your first question: 
> By the way, are those the only 3 issues you are facing at Airbnb or are there a bunch more? 

There are a few more issues we've come across internally and we've been trying to address some of them using eslint rules. 

- We want to restrict the format of arguments to either all be in one line or each argument on its own line by itself, but not mixed. [Here's a playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAEnADYkQDqEATiQCYAUAlPsADpT74D03+AhmQHUA5ugC2CGAGdhcfDiglFcDlwxZcBVBAj4G-UQEYANMJEAmM4ZEBmFsAC+atTz6CS5iVNkEYACzgcanwIAHdOZShVTnwAQTIAOjD-fngANzhqBgBybh0IbhxpAH1SHBEcACMSOBymRICEBlcuMECwAGs4qDoAUSh+GrgAYQ8q-i6zXlDOaXQwMDhpaVb9YHxpGDT0aQAVOAAPGDNqZYAHaGk4A+P8RxYAXgA+VjWuM5h0ajmISQDFCI1o4TK4mABuVyuGYeLySWC+ThNYKhCL4KIxLgAEmSqQyWVy+V0RVK5Uqw3qjUCUBasTaHW6vQGQ1qYzIEymblmmwWSxWawYGy2O32RxO+DO0kuUGutxg9yer3YdK4ErgXx+mz+6v8gOBoNiEKhsRhQhs3gR3ORIXCkRUrhKuLScEy2TyBRJZWU5NqlKaNLW7TgXR6-UGwzZJA5nWmfAI80Wy1WKsFm22X1Fx1OFyuNzFCvwLzeKo+6u+v3+uqgQJVILBkNi0L4CU85vhMjkVsCKNt6JUCiRutkGmweE4VXQ8qb+AACp9cFl8BPxOdZE18Khgls4VJ8AADIMhpnh1njSadPf4GB6KryRQY1yoOAAK34h0SInVACkAMoAeQAOUJD1ii9Cpql9Bp-VpVV8EPRkwxZUYz05GZ415JMBSFdNdjlbMpVzOUCyLZVYK4GYABFiiQq89HOM5MlgNVzjgZ06DgjxpESd58DoajhgAgBVAANH9vkyABPZgG1gh4G0cEATBACBzi0GVkFAfgtmQVBBGuJSqmoc91R-c5JkBZAYGodA4CUugIDAHS9NskBn2kQ4ACEjK6Ez+EkAAZFQnJIfTlMnc5JwsYLQrM6hrmoZAQAmKoJPIKBFJAejFBgCgcDoAJkAADgABiU65xBwSzrJc6RAVqABFdAIHgaKXO2Kpcvy-xkCsEArP4HAohEEY-nEfhEo8DLdhuIZpFaxxHCAA) illustrating this. Looks like methods that start with a capital letter or with $ and _  are treated differently. 

Would it be possible to make the list of exceptions like `_` and `$` configurable? That way we can add `fejax` and other cases to that list.

- We've also discussed maybe having a way to turn off Prettier's behavior of collapsing multiline statements. This isn't as big of an issue, but in general multiline arguments, object properties, jsx props, and named imports are intentional choices made by a developer and they aren't necessarily *prettier* one way or the other lol. 

as for this question: 
> Also, how often do those 3 happen in practice on your codebase? 

This is going to take some digging on my part so I'll get back to you with a more accurate answer soon. But afaik, this particular issue with method chaining is only affecting `fejax`, `Promise.resolve` or `Promise.all` and `window.foo` (which you've already addressed).

**Comment by vjeux** [comment]
Do you have examples where `Promise.resolve` or `Promise.all` do not behave the way you want?

> Would it be possible to make the list of exceptions like _ and $ configurable? That way we can add fejax and other cases to that list.

I'd love to know how many times you hit this in practice. Adding configuration for small things like this has a really big cost. I believe that printing a slightly less optimal output for those is a good trade-off.

> We want to restrict the format of arguments to either all be in one line or each argument on its own line by itself, but not mixed

This is indeed how prettier is supposed to work. Looks like you found a bug where there's a trailing comment in the first line: https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzOArAhgDwHQDmcMAUgMoDyAcgBQDkA9KhBAwJYDOA+nADZsE2AI15w6ASjwwAFghoAdKAAIlYWWADWAQSgATAKJQsIuAGEsvXkKyaANEoYNFKmsCUcYWGAFcOAFTgcGHsAJzgOAAdoDjgAoKUAX3ElAF4APiVgZxUHBiUAEU5jUSUYCCUIsIA3BBglMIi4LzhdVQteDjxslV0ikyoAVQANMm8QmoBPGnEAbmyExVnFEFsQCAiYNmjkUCwPZFQLGNWhEJsNYjIImzYoAmQYEO84Vd0IMAOjl5AMDhwAITOmkuWAAtnAADK3OCfDrfCDeGARREAJlhxxA1xCMRCyBA1iEE140BWmJCtxgAHU2LoZMgABwABlWMVBbAeT2+HFuBFEAEVvBB4Ojvp4hNTadJkCjVo8sGx+HdTBBQaCsHj2qTfLFjBwRQkEkA

> We've also discussed maybe having a way to turn off Prettier's behavior of collapsing multiline statements. This isn't as big of an issue, but in general multiline arguments, object properties, jsx props, and named imports are intentional choices made by a developer and they aren't necessarily prettier one way or the other lol.

We do it for objects, but because objects can be used to mean so many different things that it's hard to figure out why. For example `React.createClass({ ... })` should never be inlined, even if the content fits in one line: here object is being used as a class.

For the others, I actually believe that this is a bad idea. The goal of prettier, and a reason why I believe it's so successful, is that you can as a developer --not-- spend time formatting your code. You just write your code in a single ugly line, save and it's going to look good. It affects the way you --write-- code.

With eslint, you write your code the same way as before and the compiler yells at you if you made a mistake. It is a safety guarantee but you still have to spend all the time formatting your code.

In order to achieve this result, letting people have the choice between different ways of formatting something is actually harmful. Now you won't be sure if the input will be intentional or not. We're doing it for objects and it's a constant source of bug reports and people asking to get rid of it.

What would be great is if you could provide concrete examples where prettier didn't do the thing you expected and we can figure out if there's a way to make prettier print it best by default. What I found is that in practice, there's very often one way that people are generally happy with: maybe not the way they would actually write it in practice, but one that doesn't trigger a "omg, never in my codebase" reaction.

**Comment by vjeux** [comment]
For the latest playground bug, turns out it was fixed a week ago, before you reported it :) https://github.com/prettier/prettier/pull/3079

**Comment by sharmilajesupaul** [comment]
Oh wow, you guys move so fast! I should have checked on that bug before reporting it, my apologies.

Your sentiments on having the option to never collapse make sense. However, in regards to objects, It would be great if Prettier allowed object destructuring to be multiline. [This is the behavior Prettier exhibits now](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACSUDONXADpSpFgAWAlgDYAmATggDQFGoBmAhlXACICuNbMMtEaEi2AJ4U4mEagC+qALyoY5TADoADjQibMAbhD0QuwdEzJQbbMnYVMcYwCN+YANZwYAZU1swZKABzZBgaHkcQKggwWzZ7CIArTAAPACFXD282AFs4ABkAuFj44wgeGE1ygCZih2NfGgcaZBAnNidJaCMQbQCYAHUyKlVkAA4ABmMHbLIQsIjMAMCpAEUeCHhaiJh2weGSZCrjULZKJYBhCGzstha4im6eBwAVdoskOwc5OSA) and this destructuring pattern is used heavily throughout our codebase.

Also, `Promise.resolve` is behaving as expected in the playground (`.then` is being chained on the same line). However, I would also expect `Promise.all` to behave in the same way, but it doesn't ([here's a link illustrating this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAFAThAtgSwM5wB0AhgDakAUADprgXgJSEwAWCFABBenHldAQYcAvAD4OwADpQOHAPRyOAEwgc82OKxxQA5tIC+TMMRhgWFCkLETpshctXqsmltr1RDAbmnSM2fEQ86qQAbnAUAIJUOIQ6mhQAruikDEys7Ny8-FCCIuJSMvKKKmoaWroGDN5Q0iAANCAQVDA4AsigxHgwyABmZAQNAEboxGAA1poAylSjbsgw6AlwDSpgvf3LIABWeAAeAEIj41PEzgAy2nDrpAONCTBU9wBM17cz6AToyCCDxIMAnqRoPUQDRtDAAOo4JSsZAADgADA0CLh5otNng3KQ4ABFBIQeCvTYwP5QmEsZBPBoLYg4UhuADC2CwxG+UGgVwaCQIABU-ngifp9EA)). 

Re `fejax`: I did some digging and it is being used extensively in our codebase (about ~1800 times in ~400 files). It was introduced to be a  replacement for `$.ajax` and afaik we have no plans of removing it anytime soon. I'm not too sure how we can get around this case without being able to explicitly configure it to match the behavior of `$` and `_`. If you or @ljharb have any ideas around this I would love to hear it 🙂

**Comment by vjeux** [comment]
For Promise.resolve/all, they seem to behave the exact same way:

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAFAThAtgSwM5wB0AhgDakAUADprgXgJSEwAWCFABBenHldAQYcAvAD4OwADpQOHAPRyOAEwgc82OKxxQA5tIC+TMMRhgWFCkLETpshctXqsmltr1RDAbmnSM2fEQ86qQAbnAUAIJUOIQ6mhQAruikDEys7Ny8-FCCIuJSMvKKKmoaWroGRiZmFlb5tkUOpc7l7l4+UCAANCAQVDA4AsigxHgwyABmZAQ9AEboxGAA1poAylSLbsgw6AlwPSpgk9P7IABWeAAeAEILy2vEzgAy2nDHpDO9CTBU3wBM70+G3QBHQyBAs2IswAnqRoN0QDRtDAAOo4JSsZAADgADD0CLhtrtTng3KQ4ABFBIQeCA04wKFojEsZB-Ho7Yg4UhuADC2CwxHBUGgbx6CQIABUoXg6fp9EA

**Comment by vjeux** [comment]
Could you open another issue for object destructuring to be multiline. So we can discuss what the best solution is for this exact use case?

**Comment by lydell** [comment]
@sharmilajesupaul @vjeux The destructuring thing is already tracked in #2550

**Comment by sharmilajesupaul** [comment]
Thanks for all the fixes so far! 

@vjeux, my mistake, I think the way Prettier formats `Promise.all` & `Promise.resolve` is fine the way it is.

**Comment by myklt** [comment]
There are many popular patterns chaining method calls using enzyme. Please find below an example when Prettier creates new lines for each chained method of wrapper. Would be really nice to have an option to inline at least the first call independent on the heuristics mentioned above.

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLGAKA5AczjAZwAIYAnVHPUuAEyICMBPIgWwgFcC4F5SsAaIhgCURALwA+IsAA6UIkUhQCMIgHdSAQwAO2uKXFECAC00AbMxDUYAPAGFzCGptIARTcwD0E4QG45Cho6eqQAdDSoAG5wIqEAZqhQNBgAQuwwMNDCodqkENoi-lAAvn5yIPwg+TCo0ATIoJoqyHHmXJX0WmAA1vgAytqaYIk4yGTscJU0EGAtbZMgAFYEAB4pXb0wfZoscAAyiXBzZu1V6drpAEzHp4OkXKTIIPSaTJZQFSC5iTAA6qg0GDGZAADgADJUuCxUGNSBNISMzHAAIrsCDwG4LGCvf6A4FIS6VMiaVBmEZ2CAsFiaJ5QaBHSqcOAAFVe9SQrROcGKxSAA

**Comment by j-f1** [comment]
How does #4765 look?

**Comment by ljharb** [comment]
@j-f1 #4765 seems great - it's saying "instead of always expanding to multiple lines, prefer one-line when it was originally one-line, where possible".

However, what we want much more than that is "instead of always collapsing to one line when it fits, prefer multi-line when it was originally multi-line".

Both seem possible (ie, they don't conflict with each other). Is there any chance of also getting the second behavior?

**Comment by hwalinga** [comment]
Any update on: 

> However, what we want much more than that is "instead of always collapsing to one line when it fits, prefer multi-line when it was originally multi-line".

**Comment by thorn0** [comment]
fixed by #6685, not yet released

**Comment by randomhash** [comment]
is it broken again?
